quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,912 @@
+            min_by_key(original, new, size)
+        }
+
+        // Division "associative" (not really), left: (a / b) / c = a / (b * c)
+        (
+            ref left @ Expression::Infix(InfixExpression {
+                left: ref a,
+                operator: InfixOperator::Slash,
+                right: ref b,
+            }),
+            InfixOperator::Slash,
+            ref c,
+        ) => {
+            let original = div!(left.clone(), c.clone());
+            let bc = simplify_infix(b, InfixOperator::Star, c, limit.saturating_sub(1));
+            let new = simplify_infix(a, InfixOperator::Slash, &bc, limit.saturating_sub(1));
+            min_by_key(original, new, size)
+        }
+
+        // Right distribution: a * (b + c) = (a * b) + (a * c)
+        (
+            ref a,
+            InfixOperator::Star,
+            ref right @ Expression::Infix(InfixExpression {
+                left: ref b,
+                operator: InfixOperator::Plus,
+                right: ref c,
+            }),
+        ) => {
+            let original = mul!(a.clone(), right.clone());
+            let ab = simplify_infix(a, InfixOperator::Star, b, limit.saturating_sub(1));
+            let ac = simplify_infix(a, InfixOperator::Star, c, limit.saturating_sub(1));
+            let new = simplify_infix(&ab, InfixOperator::Plus, &ac, limit.saturating_sub(1));
+            min_by_key(original, new, size)
+        }
+
+        // Left distribution: (a + b) * c = (a * c) + (a * b)
+        (
+            ref left @ Expression::Infix(InfixExpression {
+                left: ref a,
+                operator: InfixOperator::Plus,
+                right: ref b,
+            }),
+            InfixOperator::Star,
+            ref c,
+        ) => {
+            let original = mul!(left.clone(), c.clone());
+            let ac = simplify_infix(a, InfixOperator::Star, c, limit.saturating_sub(1));
+            let bc = simplify_infix(b, InfixOperator::Star, c, limit.saturating_sub(1));
+            let new = simplify_infix(&ac, InfixOperator::Plus, &bc, limit.saturating_sub(1));
+            min_by_key(original, new, size)
+        }
+
+        //----------------------------------------------------------------
+        // Finally: other parenthesis manipulation
+        //----------------------------------------------------------------
+
+        // Mul inside Div on left with cancellation
+        (
+            Expression::Infix(InfixExpression {
+                left: ref same_1,
+                operator: InfixOperator::Star,
+                right: ref other,
+            }),
+            InfixOperator::Slash,
+            ref same_2,
+        )
+        | (
+            Expression::Infix(InfixExpression {
+                left: ref other,
+                operator: InfixOperator::Star,
+                right: ref same_1,
+            }),
+            InfixOperator::Slash,
+            ref same_2,
+        ) if **same_1 == *same_2 => *other.clone(),
```

Comment:
I think, if `limit==2` here, then `other` will not have been simplified before now because it's been further down the tree at `limit==0`.  But now we've just moved it up in the tree so we could, if we wish, call `simplify(other)` before returning it and occasionally squeak out some additional simplification.  

Similar logic, if it is valid, applies to nearby cases.

Reply:
Solid point! I’ll simplify the “return one of the things you have” cases.